### PR TITLE
feat: adds support for wildcards in redirect URIs

### PR DESCRIPTION
**Pull request by milanvanschaik**
## What kind of change does this PR introduce?
Feature: adds support for wildcards in redirect URIs.

## What is the current behavior?
Currently, it's only possible to pass exact redirect URIs in the allowed redirect URIs list. This becomes a problem when, for example, working with automatic feature branch deployments on subdomains, because currently you would have to add the new generated subdomain to the allowed redirect URIs list by hand. As discussed here: https://github.com/supabase/supabase/issues/4808

## What is the new behavior?
This PR adds support for wildcards in redirect URIs trough globbing. Now it's possible to add `*.mydomain.com` as a redirect URI and have `x.mydomain.com` and `y.mydomain.com` accepted.

Since this is quite a critical (security wise) part of GoTrue I would love to hear thoughts on this.
Also I haven't touched Go for bigger projects in 5+ years, so with that in mind, very open for feedback and suggestions. There might be a better performing implementation?

Please let me know if you have any additional questions.

## Additional context
Some things that still might need some discussion: https://github.com/supabase/supabase/issues/4808#issuecomment-1006300134

### CODE REVIEW

File: api/helpers.go [hunk 1]
```diff
@@ -125,7 +126,9 @@ func isRedirectURLValid(config *conf.Configuration, redirectURL string) bool {
 
 	// For case when user came from mobile app or other permitted resource - redirect back
 	for _, uri := range config.URIAllowList {
-		if redirectURL == uri {
+		g := glob.MustCompile(uri)
```
Comment (milanvanschaik) [inline api/helpers.go:129]: This would of course perform better if we can reuse the compiled, but not sure about the implementation (yet).

File: api/helpers.go [hunk 2]
```diff
@@ -125,7 +126,9 @@ func isRedirectURLValid(config *conf.Configuration, redirectURL string) bool {
...
+
+		if g.Match(redirectURL) {
```
Comment (milanvanschaik) [inline api/helpers.go:131]: Could be
```
return g.Match(redirectURL)
```
but I like it explicit for readability

**Comment by awalias** [comment]
alternative solution: allow programmatic manipulation of the redirect URIs https://github.com/supabase/gotrue/issues/296

**Review by J0 — COMMENTED** [review]
Looks good to me. We can discuss further during catchup if  needed

**Review by J0 — APPROVED** [review]

**Review by awalias — APPROVED** [review]
looks good but can we add a couple more test cases:

`https://*.x.com` where `https://a.b.x.com` fails
and
`https://*.*.x.com` where `https://a.b.x.com` passes

and maybe a few more if we can get creative to try and catch all the edge cases

**Comment by bnjmnt4n** [comment]
I think it would be great to link to the documentation for valid glob syntax (https://pkg.go.dev/github.com/gobwas/glob#Compile).

The docs also recommend doing compilation of globs a single time for performance. I'm guessing this might not be an issue for most users, but eventually repeated compilation of globs on every redirect request might lead to performance penalties.

**Comment by kangmingtay** [comment]
> The docs also recommend doing compilation of globs a single time for performance. I'm guessing this might not be an issue for most users, but eventually repeated compilation of globs on every redirect request might lead to performance penalties.

Thanks for catching this! I've moved the glob compilation for the uri allowed list to when gotrue initially starts up and loads the config so we only compile the globs once.

**Review by kangmingtay — APPROVED** [review]

File: api/helpers.go (hunk 1 above)
Comment (kangmingtay) [reply api/helpers.go]: i've moved it to compile when gotrue starts [here](https://github.com/supabase/gotrue/pull/334/files#diff-5b7f9f2790a5795a69654b2ba8c72112da865800efa1110b181c6c70de96c9dbR346-R352) and loads the config. This would allow use to reuse the compiled globs on runtime.

**Comment by mrpbennett** [comment]
This is awesome!! 

I am having a slight issue though hoping someone could help.

Screenshot 2022-05-13 at 16 37 42

After clicking the button which fires the login I get redirected to `https://s3-directory-npi.vercel.app/#` 

This is the function 

```javascript
const signInWithGoogle = async () => {
    const {user, session, error} = await supabase.auth.signIn(
      {
        provider: 'google',
      },
      {
        redirectTo: 'https://s3-directory-npi.vercel.app/home',
      },
    )

    if (error) {
      popupValidation('error', error.message)
    }
  }
```

any ideas would be great.

**Comment by pie6k** [comment]
Just lost like 2 hours trying to solve this issue not knowing about that.

Please try to add notes about such a feature in the panel in the future. Simple "You can use glob batterns" would do the job. Copy in the panel clearly indicates I need strict, exact match urls

[[rage level / trying to write politely]] ratio = 8000

CleanShot 2022-09-07 at 11 30 49@2x

**Comment by Jutanium** [comment]
Seconding. Please update the copy on the dashboard!

**Comment by thorwebdev** [comment]
Until we have docs for this you can refer to https://github.com/supabase/supabase/issues/4808#issuecomment-1148467741 for an overview.
